browser/de_amp/de_amp_browsertest.cc

```diff
@@ -0,0 +1,278 @@
+#include "chrome/browser/ui/browser_commands.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "components/keep_alive_registry/keep_alive_types.h"
+#include "components/keep_alive_registry/scoped_keep_alive.h"
+#include "components/prefs/testing_pref_service.h"
+#include "content/public/test/browser_test.h"
+#include "content/public/test/content_mock_cert_verifier.h"
+#include "content/public/test/test_navigation_observer.h"
+#include "net/dns/mock_host_resolver.h"
+#include "net/test/embedded_test_server/embedded_test_server.h"
+#include "third_party/re2/src/re2/re2.h"
+
+#if defined(OS_ANDROID)
+#include "chrome/test/base/android/android_browser_test.h"
+#else
+#include "chrome/test/base/in_process_browser_test.h"
+#include "chrome/test/base/ui_test_utils.h"
+#endif
+
+const char kTestHost[] = "a.test.com";
+const char kTestAmpPage[] = "/test.html";
+const char kTestSimpleNonAmpPage[] = "/simple.html";
+const char kTestCanonicalPage[] = "/simple_canonical.html";
+const char kTestBody[] =
+    "<html amp>\n"
+    "<head>\n"
+    "<link rel='canonical' "
+    "href='https://%s:%s%s'>\n"
+    "</head></html>";
+class DeAmpBrowserTest : public PlatformBrowserTest {
+ public:
+  DeAmpBrowserTest() {
+    feature_list_.InitAndEnableFeature(de_amp::features::kBraveDeAMP);
+  }
+
+  void SetUpOnMainThread() override {
+    mock_cert_verifier_.mock_cert_verifier()->set_default_result(net::OK);
+    host_resolver()->AddRule("*", "127.0.0.1");
+    brave::RegisterPathProvider();
+    base::FilePath test_data_dir;
+    base::PathService::Get(brave::DIR_TEST_DATA, &test_data_dir);
+    https_server_.reset(new net::EmbeddedTestServer(
+        net::test_server::EmbeddedTestServer::TYPE_HTTPS));
+
+    https_server_->ServeFilesFromDirectory(test_data_dir);
+    content::SetupCrossSiteRedirector(https_server_.get());
+    InProcessBrowserTest::SetUpOnMainThread();
+  }
+
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    InProcessBrowserTest::SetUpCommandLine(command_line);
+    mock_cert_verifier_.SetUpCommandLine(command_line);
+  }
+
+  void SetUpInProcessBrowserTestFixture() override {
+    InProcessBrowserTest::SetUpInProcessBrowserTestFixture();
+    mock_cert_verifier_.SetUpInProcessBrowserTestFixture();
+  }
+
+  void TearDownInProcessBrowserTestFixture() override {
+    InProcessBrowserTest::TearDownInProcessBrowserTestFixture();
+    mock_cert_verifier_.TearDownInProcessBrowserTestFixture();
+  }
+
+  de_amp::DeAmpService* service() {
+    return de_amp::DeAmpServiceFactory::GetForBrowserContext(
+        web_contents()->GetBrowserContext());
+  }
+
+  content::WebContents* web_contents() {
+    return browser()->tab_strip_model()->GetActiveWebContents();
+  }
+
+  void TogglePref(const bool on) {
+    service()->ToggleDeAmp(on);
+    web_contents()->GetController().Reload(content::ReloadType::NORMAL, false);
+  }
+
+  void NavigateToURLAndWaitForRedirects(const GURL& original_url,
```

Comment:
I suggest return `landing_url` here and them check the result using:
`EXPECT_EQ(NavigateToURLAndWaitForLangingUrl(some_url), landing_url);`

Reply:
ui_test_utils::UrlLoadObserver needs the URL to wait for though